### PR TITLE
Add error handling/boundaries for plugins

### PR DESCRIPTION
**Pull request by illume**
# Add error handling/boundaries for plugins

Stopped plugins with errors crashing Headlamp:
- On plugin load (script syntax error, or other error)
- When components are rendering

Extracted plugin loading and executing into a Plugins component (moving the rest of it there will come in a future PR).

Wrapped `<ErrorBoundary>` around where plugins can insert components into the app.

Fixes https://github.com/kinvolk/headlamp/issues/597

## How to Test

Introduce some errors into the example plugins.

- [x] Add an error into a plugin component (`throw 'error';`) and Headlamp should still load.
- [x] Add an error into a compiled plugin, and Headlamp should still load.

```tsx
import { registerAppBarAction } from '@kinvolk/headlamp-plugin/lib';

registerAppBarAction(() => {
  throw 'meow';
});
```

### CODE REVIEW

File: frontend/src/components/common/ErrorBoundary/ErrorBoundary.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ */
+export default class ErrorBoundary extends Component<ErrorBoundaryProps, State> {
+  state: State = { error: null };
+  static getDerivedStateFromError(error: Error | null) {
+    return { error };
+  }
+
+  //   componentDidCatch(error, errorInfo) {
+  //     logError(error, errorInfo);
+  //   }
```
Comment (joaquimrocha) [inline frontend/src/components/common/ErrorBoundary/ErrorBoundary.tsx:44]: Maybe unintendedly left?

File: frontend/src/redux/actions/actions.tsx [hunk 2]
```diff
@@ -77,6 +77,7 @@ export interface Action {
 type SidebarType = UIState['sidebar'];
 
 export type HeaderActionType = ((...args: any[]) => JSX.Element | null) | null | ReactElement;
+// export type HeaderActionType = React.FunctionComponent | null | ReactElement;
```
Comment (joaquimrocha) [inline frontend/src/redux/actions/actions.tsx:80]: Is this intended?

File: frontend/src/plugin/Plugins.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        (function (str: string) {
+          try {
+            const result = eval(str);
+            return result;
+          } catch (e) {
+            console.error(e);
+          }
+        }.call({}, source));
+      });
+      await initializePlugins();
```
Comment (joaquimrocha) [inline frontend/src/plugin/Plugins.tsx:44]: We're calling `store.dispatch({ type: UI_RESET_PLUGIN_VIEWS });` within this method which runs after loading the external plugins AFAICS. Won't this be a problem now that the plugins perform code right when they're loaded (before it worked on the assumption we'd load the code but only initialize the plugins after resetting views, etc.)?

File: frontend/src/plugin/Plugins.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+      sources.forEach(source => {
+        (function (str: string) {
+          try {
+            const result = eval(str);
+            return result;
+          } catch (e) {
+            console.error(e);
+          }
+        }.call({}, source));
```
Comment (joaquimrocha) [inline frontend/src/plugin/Plugins.tsx:42]: I'm surprise this works fine like this. Before we were injecting the scripts into the document. Not sure how things differ or what's best, just curious.

File: frontend/src/plugin/Plugins.tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        )
+      );
+
+      sources.forEach(source => {
+        (function (str: string) {
+          try {
+            const result = eval(str);
+            return result;
+          } catch (e) {
+            console.error(e);
```
Comment (joaquimrocha) [inline frontend/src/plugin/Plugins.tsx:40]: Maybe prefix the error so we have that context when we see it?

File: frontend/src/plugin/Plugins.tsx [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    const fetchData = async () => {
+      const pluginsScriptPaths = (await fetch(`${helpers.getAppUrl()}plugins/list`).then(resp =>
+        resp.json()
+      )) as string[];
+
+      const sources = await Promise.all(
+        pluginsScriptPaths.map(path =>
+          fetch(`${helpers.getAppUrl()}${path}`).then(resp => resp.text())
+        )
+      );
```
Comment (joaquimrocha) [inline frontend/src/plugin/Plugins.tsx:30]: Nice! I guess we get some speed improvements by having these (sort of) parallel.

File: frontend/src/components/common/index.test.ts [hunk 7]
```diff
@@ -12,6 +12,7 @@ const avoidCheck = [
   'AlertNotification',
   'ConfirmButton',
   'NamespacesAutocomplete',
+  'ErrorBoundary',
```
Comment (joaquimrocha) [inline frontend/src/components/common/index.test.ts:15]: 👍

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
This is pretty cool! I left a couple of comments.
As a second iteration, maybe we could have a mode where we use a visible and adequate fallback, so users can see at least what plugin has generated the errors and where.

Also, maybe on the plugin defined routes' components we should always show a fallback component with an error message?

File: frontend/src/plugin/Plugins.tsx (hunk 3 above)
Comment (illume) [reply frontend/src/plugin/Plugins.tsx:53]: Thanks. I moved this so the initialization is done first before loading the plugins.

**Review by illume — COMMENTED** [review]
> Hrmm. Github changed so my replies to your comments are now done with a review, and all my comments were pending and not sent right away. Kind of strange... but ok github.

I've addressed all your comments but left them unresolved so you can see.

File: frontend/src/components/common/ErrorBoundary/ErrorBoundary.tsx (hunk 1 above)
Comment (illume) [reply frontend/src/components/common/ErrorBoundary/ErrorBoundary.tsx]: Thanks. Removed it.

(Was leaving it there for a later, but it's easy to find in the docs so removed it.)

**Review by joaquimrocha — APPROVED** [review]

**Pull request merged**